### PR TITLE
3126 refactor config registry

### LINKED ISSUE #3126: Refactor app configurations sync to use download managers
**Describe the feature request.**
We need to refactor the current configuration sync to move from custom server side calls to using the FHIR SDK DownloadManagers to fetch configurations. The current approach requires a custom engine library implementation that duplicates this existing functionality.

**Additional context**
If available add any other context or screenshots about the feature here.

**Acceptance criteria**
Configuration sync should function correctly. This can be tested by entering the app id and logging in to the app with the ability to use config resources e.g. Launching Questionnaires and performing a success saving (confirms data extraction works correctly since StructureMaps were successfully synced)

**Area path**
A list of ordered steps in the app on usage of the feature to support anyone testing it e.g. Code reviewer, QA e.g.
1. Launch the app
2. Enter APP ID (should proceed to Login Page)
3. Enter Credentials and login
4. Wait for sync to complete
5. Add a client e.g. Patient or Household. Form save should complete successfully

**Implementation plan (For Engineers)**
The plan for implementing the solution e.g. via a _description_ or a _check list_ for the various ordered tasks that will need to be completed.
_i.e. Describe how you intend to solve the problem_

- Refactor the ConfigRegistry class and integrate suitable DownloadManagers provided by the FHIR SDK

### PR DESCRIPTION
**Pull request by AbdulWahabMemon**
**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Fixes #3126

**Engineer Checklist**
- [ ] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [ ] I have added any strings visible on UI components to the `strings.xml` file
- [ ] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [ ] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [x] I have built and run the FHIRCore app to verify my change fixes the issue and/or does not break the app 
- [ ] I have checked that this PR does NOT introduce **breaking changes** that require an update to **_Content_** and/or **_Configs_**? _If it does add a sample here or a link to exactly what changes need to be made to the content._

**Code Reviewer Checklist**
- [ ] I have verified **Unit tests** have been written for any new feature(s) and edge cases
- [ ] I have verified any strings visible on UI components are in the `strings.xml` file
- [ ] I have verifed the [CHANGELOG.md](./CHANGELOG.md) file has any notable changes to the codebase
- [ ] I have verified the solution has been implemented in a configurable and generic way for reuseable components
- [ ] I have built and run the FHIRCore app to verify the change fixes the issue and/or does not break the app

### CODE REVIEW

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt [hunk 1]
```diff
@@ -830,4 +898,276 @@ constructor(
         ResourceType.Parameters,
       )
   }
+
+  private suspend fun fetchCompositionConfig(coroutineScope: CoroutineScope) {
+    val compConfigFlow1 = Sync.oneTimeSync<CompositionSyncWorker>(context)
+    compConfigFlow1.handleCompositionSyncJobStatus(coroutineScope)
+    compConfigFlow1.collect { compConfigFlowJobStatus ->
+      when (compConfigFlowJobStatus) {
```
Comment (ndegwamartin) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt:906]: SAB(Same as above). `compositionConfigFlowJobStatus`

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt [hunk 2]
```diff
@@ -830,4 +898,276 @@ constructor(
...
+            }
+            saveSyncSharedPreferences(patientRelatedResourceTypes.toList())
+        }
+      }
+    }
+  }
+
+  private suspend fun fetchCompositionManifestRequest(coroutineScope: CoroutineScope) {
+
+    val compositionManifestFlow4 = Sync.oneTimeSync<CompositionManifestSyncWorker>(context)
```
Comment (ndegwamartin) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt:1038]: we can remove the No. 4 suffix here since it is the only flow in this methods scope

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt [hunk 3]
```diff
@@ -830,4 +898,276 @@ constructor(
...
+          Timber.d("configSync composition Manifest FlowJobStatus succeeded")
+        } else -> {
+        // do nothing Timber.d("#### compositionManifestFlow4 JobStatus other than succeeded")
+        }
+      }
+    }
+  }
+
+  private suspend fun fetchCompositionListRequest(coroutineScope: CoroutineScope) {
+    val compListFlow2 = Sync.oneTimeSync<CompositionListSyncWorker>(context)
```
Comment (ndegwamartin) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt:1052]: SAB

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt [hunk 4]
```diff
@@ -830,4 +898,276 @@ constructor(
...
+          }
+        } else -> {
+        // do nothing Timber.d("#### compListJobStatus other than succeeded ")
+        }
+      }
+    }
+  }
+
+  private suspend fun fetchCompositionConfigRequest(coroutineScope: CoroutineScope) {
+    val compositionConfigFlow3 = Sync.oneTimeSync<CompositionConfigSyncWorker>(context)
```
Comment (ndegwamartin) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt:1071]: SAB

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt [hunk 5]
```diff
@@ -830,4 +898,276 @@ constructor(
         ResourceType.Parameters,
       )
   }
+
+  private suspend fun fetchCompositionConfig(coroutineScope: CoroutineScope) {
+    val compConfigFlow1 = Sync.oneTimeSync<CompositionSyncWorker>(context)
```
[the remark this answers is not in the document]
Comment (AbdulWahabMemon) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt]: updated

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt (hunk 1 above)
Comment (AbdulWahabMemon) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt]: updated

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt [hunk 6]
```diff
@@ -830,4 +898,276 @@ constructor(
...
+                Timber.d("configSync process compositionContent if Not LIST block")
+                val chunkedResourceIdList = entry.value.chunked(MANIFEST_PROCESSOR_BATCH_SIZE)
+                chunkedResourceIdList.forEach { parentIt ->
+                  val resourceIds: List<String> =
+                    parentIt.map { sectionComponent -> sectionComponent.focus.extractId() }
+                  val resourceType = entry.key
+
+                  resourceIds.forEach {
+                    Timber.d("configSync composition Config adding resourceType - ${resourceType} and id= ${it}")
+                    val compConfigParamPairsReq = mutableListOf<Pair<ResourceType, Map<String, String>>>()
```
[the remark this answers is not in the document]
Comment (AbdulWahabMemon) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt]: updated

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt (hunk 2 above)
Comment (AbdulWahabMemon) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt]: updated

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt (hunk 3 above)
Comment (AbdulWahabMemon) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt]: updated

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt [hunk 7]
```diff
@@ -830,4 +898,276 @@ constructor(
...
+        }
+      }
+      Timber.d("configSync composition ListItemRequestQue size - ${syncParamSource.compConfigRequestQue.size}")
+      fetchCompListItemRequest(coroutineScope)
+    }
+  }
+
+
+  private suspend fun fetchCompListItemRequest(coroutineScope: CoroutineScope) {
+    val compListFlow2 = Sync.oneTimeSync<CompListItemSyncWorker>(context)
```
[the remark this answers is not in the document]
Comment (AbdulWahabMemon) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt]: updated

File: android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt (hunk 4 above)
Comment (AbdulWahabMemon) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/ConfigurationRegistry.kt]: updated
